core/src/main/java/jeeves/server/context/ServiceContext.java

```diff
@@ -52,21 +53,179 @@
 
 /**
  * Contains the context for a service execution.
+ *
+ * When creating a ServiceContext you are responsible for manging its use on the current thread and any cleanup:
+ * <pre><code>
+ * try {
+ *    context = serviceMan.createServiceContext("AppHandler", appContext);
+ *    context.setAsThreadLocal();
+ *    ...
+ * } finally {
+ *    context.clearAsThreadLocal();
+ *    context.clear();
+ * }</code></pre>
+ *
+ * @see ServiceManager
  */
-public class ServiceContext extends BasicContext {
+public class ServiceContext extends BasicContext implements AutoCloseable {
 
+    /**
+     * ServiceContext is managed as a thread locale using setAsThreadLocal, clearAsThreadLocal and clear methods.
+     * ThreadLocalPolicy defines the behaviour of these methods double checking that they are being used correctly.
+     */
+    public static enum ThreadLocalPolicy {
+        /** Direct management of thread local with no checking. */
+        DIRECT,
+        /** Check behavior and log any unexpected use. */
+        TRACE,
+        /** Raise any {@link IllegalStateException} for unexpected behaviour */
+        STRICT };
+    /**
+     * Use -Djeeves.server.context.service.policy to define policy:
+     * <ul>
+     *     <li>direct: direct management of thread local with no checking</li>
+     *     <li>trace: check behavior and log unusal use</li>
+     *     <li>strict: raise illegal state exception for unusual behavior</li>
+     * </ul>
+     */
+    private static final ThreadLocalPolicy POLICY;
+    static {
+        String property = System.getProperty("jeeves.server.context.policy", "TRACE");
```

Comment:
I was only willing to default to TRACE for maintenance branch, on main we could consider defaulting to STRICT (at least for local development?).